src/core/ComboBox/ComboBox.tsx

```diff
@@ -380,84 +446,309 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
+    const refs = useMergedRefs(inputRef, forwardedRef);
+
+    const focusedIndexRef = React.useRef(focusedIndex ?? -1);
+    React.useEffect(() => {
+      focusedIndexRef.current = focusedIndex ?? -1;
+    }, [focusedIndex]);
+
+    const getIdFromIndex = (index: number) => {
+      return (
+        menuRef.current?.querySelector(`[data-iui-index="${index}"]`)?.id ?? ''
+      );
+    };
+
+    const handleKeyDown = React.useCallback(
+      (event: React.KeyboardEvent<HTMLInputElement>) => {
+        (() => {
+          const length =
+            menuRef.current?.querySelectorAll('[data-iui-index]').length ?? 0;
+
+          if (event.key === 'ArrowDown') {
+            event.preventDefault();
+            if (!isOpen) {
+              return dispatch(['open']);
             }
-          }}
-          animation='shift-away'
-          duration={200}
-          {...dropdownMenuProps}
-          content={
-            <Menu
-              id={`${id}-list`}
-              className='iui-scroll'
-              style={{
-                minWidth,
-                maxWidth: `min(${minWidth * 2}px, 90vw)`,
-                maxHeight: 300,
-              }}
-              setFocus={false}
-              role='listbox'
-              ref={menuRef}
-            >
-              {menuItems}
-            </Menu>
-          }
-          onHide={(instance) => {
-            const selectedIndex = options.findIndex(
-              ({ value }) => value === selectedValue,
-            );
-            setFocusedIndex(selectedIndex);
-            if (selectedIndex > -1) {
-              setInputValue(options[selectedIndex].label); // update input value to be same as selected value
+
+            if (focusedIndexRef.current === -1 && length > 0) {
+              return dispatch(['focus', 0]);
             }
-            dropdownMenuProps?.onHide?.(instance);
-          }}
-        >
-          <Input
-            ref={inputRef}
-            onKeyDown={onKeyDown}
-            onFocus={() => setIsOpen(true)}
-            value={inputValue}
-            aria-activedescendant={
-              isOpen && focusedIndex > -1
-                ? getOptionId(focusedIndex)
-                : undefined
+
+            if (length === 0) {
+              return;
             }
-            role='combobox'
-            aria-controls={isOpen ? `${id}-list` : undefined}
-            aria-autocomplete='list'
-            spellCheck={false}
-            autoCapitalize='none'
-            autoCorrect='off'
-            {...inputProps}
-            onChange={onInput}
-          />
-        </Popover>
-        <span
-          ref={toggleButtonRef}
-          className={cx('iui-end-icon', {
-            'iui-actionable': !inputProps?.disabled,
-            'iui-disabled': inputProps?.disabled,
-            'iui-open': isOpen,
-          })}
-          onClick={() => {
+
+            while (true) {
+              const currentElement =
+                menuRef.current?.querySelector(
```

Comment:
this is outdated, please look at https://github.com/iTwin/iTwinUI-react/pull/623/commits/69056c9770d96b6bdc0233ded4ce3fca26e384da

but it's still querying dom. can you elaborate on the map part and how it would work with custom items?